src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -273,6 +273,36 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let result = Scalar::from_uint(truncated_bits, layout.size);
                 self.write_scalar(result, dest)?;
             }
+            sym::offset | sym::arith_offset => {
+                let offset_count = self.read_scalar(args[1])?.to_machine_isize(self)?;
+                let ptr = self.read_scalar(args[0])?.not_undef()?;
+                let check_bounds = intrinsic_name == sym::offset;
+                // offset = offset_count * size_of::<T>()
+                let pointee_ty = substs.type_at(0);
+                let pointee_size = i64::try_from(self.layout_of(pointee_ty)?.size.bytes()).unwrap();
+                let (offset, overflowed) = offset_count.overflowing_mul(pointee_size);
```

Comment:
Better use `checked_mul`, which has no risk of accidentally using the result when there was an overflow.

Reply:
Fixed by separating the implementations for `offset` and `arith_offset`.